### ISSUE TITLE
[2.0.1] Kuberenetes Deployment - Segfault at 6 ip 0000561a6f9d0ae0

### ISSUE THREAD

**Issue by DandyDeveloper** [haproxy/haproxy#159]
## Output of `haproxy -vv` and `uname -a`
```
HA-Proxy version 2.0.1 2019/06/26 - https://haproxy.org/
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = gcc
  CFLAGS  = -O2 -g -fno-strict-aliasing -Wdeclaration-after-statement -fwrapv -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-old-style-declaration -Wno-ignored-qualifiers -Wno-clobbered -Wno-missing-field-initializers -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_GETADDRINFO=1 USE_OPENSSL=1 USE_LUA=1 USE_ZLIB=1

Feature list : +EPOLL -KQUEUE -MY_EPOLL -MY_SPLICE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED -REGPARM -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H -VSYSCALL +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 -MY_ACCEPT4 +ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL -SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=8).
Built with OpenSSL version : OpenSSL 1.1.0j  20 Nov 2018
Running on OpenSSL version : OpenSSL 1.1.0j  20 Nov 2018
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2
Built with Lua version : Lua 5.3.3
Built with network namespace support.
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with zlib version : 1.2.8
Running on zlib version : 1.2.8
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with PCRE2 version : 10.22 2016-07-29
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with the Prometheus exporter as a service

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTX        side=FE|BE     mux=H2
              h2 : mode=HTTP       side=FE        mux=H2
       <default> : mode=HTX        side=FE|BE     mux=H1
       <default> : mode=TCP|HTTP   side=FE|BE     mux=PASS

Available services :
	prometheus-exporter

Available filters :
	[SPOE] spoe
	[COMP] compression
	[CACHE] cache
	[TRACE] trace
```
```
Linux dhash-redis-ha-haproxy-5f7679fc77-wqrf2 4.19.25-coreos #1 SMP Sat Mar 9 01:05:06 -00 2019 x86_64 GNU/Linux
```

## What's the configuration?
```
defaults REDIS
  mode tcp
  timeout connect  4s
  timeout server  30s
  timeout client  30s
# Check Sentinel and whether they are nominated master
backend check_if_redis_is_master_0
  mode tcp
  option tcp-check
  tcp-check connect
  tcp-check send PING\r\n
  tcp-check expect string +PONG
  tcp-check send SENTINEL\ get-master-addr-by-name\ mymaster\r\n
  tcp-check expect string 10.100.128.121
  tcp-check send QUIT\r\n
  tcp-check expect string +OK
  server R0 dhash-redis-ha-announce-0:26379 check inter 1s
  server R1 dhash-redis-ha-announce-1:26379 check inter 1s
  server R2 dhash-redis-ha-announce-2:26379 check inter 1s
# Check Sentinel and whether they are nominated master
backend check_if_redis_is_master_1
  mode tcp
  option tcp-check
  tcp-check connect
  tcp-check send PING\r\n
  tcp-check expect string +PONG
  tcp-check send SENTINEL\ get-master-addr-by-name\ mymaster\r\n
  tcp-check expect string 10.103.158.92
  tcp-check send QUIT\r\n
  tcp-check expect string +OK
  server R0 dhash-redis-ha-announce-0:26379 check inter 1s
  server R1 dhash-redis-ha-announce-1:26379 check inter 1s
  server R2 dhash-redis-ha-announce-2:26379 check inter 1s
# Check Sentinel and whether they are nominated master
backend check_if_redis_is_master_2
  mode tcp
  option tcp-check
  tcp-check connect
  tcp-check send PING\r\n
  tcp-check expect string +PONG
  tcp-check send SENTINEL\ get-master-addr-by-name\ mymaster\r\n
  tcp-check expect string 10.100.165.193
  tcp-check send QUIT\r\n
  tcp-check expect string +OK
  server R0 dhash-redis-ha-announce-0:26379 check inter 1s
  server R1 dhash-redis-ha-announce-1:26379 check inter 1s
  server R2 dhash-redis-ha-announce-2:26379 check inter 1s

# decide redis backend to use
frontend ft_redis
  bind *:6379
  use_backend bk_redis

# Check all redis servers to see if they think they are master
backend bk_redis
  mode tcp
  option tcp-check
  tcp-check connect
  tcp-check send PING\r\n
  tcp-check expect string +PONG
  tcp-check send info\ replication\r\n
  tcp-check expect string role:master
  tcp-check send QUIT\r\n
  tcp-check expect string +OK
  use-server R0 if { srv_is_up(R0) } { nbsrv(check_if_redis_is_master_0) ge 2 }
  server R0 dhash-redis-ha-announce-0:6379 check inter 1s fall 1 rise 1
  use-server R1 if { srv_is_up(R1) } { nbsrv(check_if_redis_is_master_1) ge 2 }
  server R1 dhash-redis-ha-announce-1:6379 check inter 1s fall 1 rise 1
  use-server R2 if { srv_is_up(R2) } { nbsrv(check_if_redis_is_master_2) ge 2 }
  server R2 dhash-redis-ha-announce-2:6379 check inter 1s fall 1 rise 1
```

## Steps to reproduce the behavior
I recently included a HAProxy inside of a HA Redis server, it's working but all of my clusters running this are restarting with this segfault error. 

1. Clone/Checkout: https://github.com/DandyDeveloper/charts/tree/stable/redis-ha/add-proxy-support-external-clients
2. Using helm, install the chart with `haproxy.enabled: true`
3. Wait an hour or two. Segfaults should be prominent.

## Actual behavior
I can't see any specific behaviours or exact time it takes to restart, it seems to be random whenever it hits whatever error it hits. 

This is the log:
```
[ALERT] 189/013924 (1) : Current worker #1 (8) exited with code 139 (Segmentation fault)
[ALERT] 189/013924 (1) : exit-on-failure: killing every processes with SIGTERM
[WARNING] 189/013924 (1) : All workers exited. Exiting... (139)
```

Segfault is always followed by this message in `dmesg`:
```
[5070988.391650] haproxy[1784]: segfault at 6 ip 000055ece28a7ae0 sp 00007efd97a35410 error 4 in haproxy[55ece27d2000+1ff000]
[5070988.402889] Code: e9 0d 00 00 48 8b 00 49 39 c6 49 89 84 24 08 01 00 00 48 89 c1 75 e3 49 8b 54 24 40 4c 89 f1 4c 89 f0 e9 5f fe ff ff 45 31 ed <41> f6 47 06 10 75 0a 48 8b 04 24 f6 40 05 01 74 0c 31 d2 31 f6 4c
```

## Expected behavior
No segfault.

## Do you have any idea what may have caused this?
It's possible my interval / check logic is causing problems? We always have atleast 2 servers down at the time when this is running as they are slaves. 

## Do you have an idea how to solve the issue?
No.

**Comment by wtarreau** [comment]
Nothing in your config should be responsible for a segfault, you have hit
a serious bug. However it is impossible to figure what is happening without
a backtrace or preferably a core file to analyze. Please add "set-dumpable"
in your global section, possibly disable chroot if you have it, and make
sure that your system allows core dumps (i.e. if /proc/sys/core/core_pattern
points to a directory, make sure it's accessible with full write permissions,
and if it references a dump service, make sure this service is actually
installed and configured).

**Comment by capflam** [comment]
I don't know if it is related, but take a look at the issue #161. You have some `tcp-check` in your configuration, so maybe you hit the same issue. Also try to get the [last 2.0 snapshot](http://www.haproxy.org/download/2.0/src/snapshot/haproxy-ss-LATEST.tar.gz). Another bug was fixed on the tcp-check. See issue #148

**Comment by DandyDeveloper** [comment]
@wtarreau Sorry I'm dealing with a few other issues, I'll get that to you if the latest snapshot doesnt' work per @capflam 

@capflam Just to confirm, is the image from 3 hours a go updated with that tcp-check fix?

**Comment by capflam** [comment]
It includes the fix about the issue #148. But for the last one, the issue #161, @cognet has proposed a [patch](https://github.com/haproxy/haproxy/files/3373815/0001-WIP-tcpcheck-segfault.txt). So you need to apply it by hand for now.

**Comment by wtarreau** [comment]
Given that 2.0.2 is now released with many fixes for such similar issues, it would be nice to recheck.

**Comment by DandyDeveloper** [comment]
@wtarreau Any way we can get the latest docker image uploaded to the official repo?

**Comment by wtarreau** [comment]
I have no idea, I don't even know who does these images nor when. The only official repo I know is the source repo : http://www.haproxy.org/download/2.0/src/

**Comment by DandyDeveloper** [comment]
@wtarreau Weird, I've made my own image then and I'll see how it performs! Will report an update tomorrow.

**Comment by wtarreau** [comment]
OK, thank you!

**Comment by TimWolla** [comment]
@DandyDeveloper The Docker image will update once this is merged: https://github.com/docker-library/official-images/pull/6306

**Comment by wtarreau** [comment]
Thanks Tim for the heads up!

**Comment by DandyDeveloper** [comment]
@wtarreau Looks fixed, overnight I've had 0 segfault crashes (infact, I've had 0 crashes)! Thanks everyone. 

@TimWolla Thank you. We've got our own for now. I'll update it soonish. 

To those returning **This is fixed in HAProxy 2.0.2**
